peer.go

```diff
@@ -1,640 +1,444 @@
-		}
+// SendWithTimeout invokes Mux.SendWithTimeout on a peers default mux session.
+func (p *Peer) SendWithTimeout(opcode byte, msg []byte, timeout time.Duration) error {
+	return p.m.SendWithTimeout(opcode, msg, timeout)
+}
 
-		b, errs := p.beforeMessageReceivedCallbacks.RunCallbacks(buf, p.node)
-		if len(errs) > 0 {
-			log.Warn().Errs("errors", errs).Msg("Got errors running BeforeMessageReceived callbacks.")
+// Recv invokes Mux.Recv on a peers default mux session.
+func (p *Peer) Recv(opcode byte) <-chan Wire {
+	return p.m.Recv(opcode)
+}
 
-			p.Disconnect()
-			continue
-		}
-		buf = b.([]byte)
+// LockOnRecv invokes Mux.LockOnRecv on a peers default mux session.
+func (p *Peer) LockOnRecv(opcode byte) Locker {
+	return p.m.LockOnRecv(opcode)
+}
 
-		opcode, msg, err := p.DecodeMessage(buf)
+func (p *Peer) AfterSend(f func()) {
+	p.afterSendLock.Lock()
+	p.afterSend = append(p.afterSend, f)
+	p.afterSendLock.Unlock()
+}
 
-		if opcode == OpcodeNil || err != nil {
-			p.onConnErrorCallbacks.RunCallbacks(p.node, errors.Wrap(err, "failed to decode message"))
+func (p *Peer) AfterRecv(f func()) {
+	p.afterRecvLock.Lock()
+	p.afterRecv = append(p.afterRecv, f)
+	p.afterRecvLock.Unlock()
+}
 
-			p.Disconnect()
-			continue
-		}
+func (p *Peer) RegisterSignal(name string) func() {
+	p.signalsLock.Lock()
+	signal, exists := p.signals[name]
+	if !exists {
+		signal = make(chan struct{})
+		p.signals[name] = signal
+	}
+	p.signalsLock.Unlock()
 
-		c, _ := p.receiveQueues.LoadOrStore(opcode, receiveHandle{hub: make(chan Message), lock: make(chan struct{}, 1)})
-		recv := c.(receiveHandle)
+	return func() {
+		var closed bool
 
 		select {
-		case recv.hub <- msg:
-			recv.lock <- struct{}{}
-			<-recv.lock
-		case <-time.After(p.node.receiveMessageTimeout):
-			p.Disconnect()
-			continue
+		case _, ok := <-signal:
+			if !ok {
+				closed = true
+			}
+		default:
 		}
 
-		if errs := p.afterMessageReceivedCallbacks.RunCallbacks(p.node); len(errs) > 0 {
-			log.Warn().Errs("errors", errs).Msg("Got errors running AfterMessageReceived callbacks.")
-
-			p.Disconnect()
-			continue
+		if !closed {
+			close(signal)
 		}
 	}
 }
 
-// SendMessage sends a message whose type is registered with Noise to a specified peer. Calling
-// this function will block the current goroutine until the message is successfully sent. In
-// order to not block, refer to `SendMessageAsync(message Message) <-chan error`.
-//
-// It is guaranteed that all messages are sent in a linearized order.
-//
-// It returns an error should it take too long to send a message, the message is not registered
-// with Noise, or there are message that are blocking the peers send worker.
-func (p *Peer) SendMessage(message Message) error {
-	payload, err := p.EncodeMessage(message)
-	if err != nil {
-		return errors.Wrap(err, "failed to serialize message contents to be sent to a peer")
+func (p *Peer) WaitFor(name string) {
+	p.signalsLock.Lock()
+	signal, exists := p.signals[name]
```

Comment:
shouldn't it be like
```golang
func (p *Peer) WaitFor(name string) {
    close = p.RegisterSignal(name)
    defer close() // not sure about this one though

    p.signalsLock.Lock()
    signal := p.signals[name]
    p.signalsLock.Unlock()

    <-signal
}
```
but getting signal from the map is kinda clunky, so it seems like `RegisterSignal` should return not only closer for channel but the channel itself as well.

Reply:
`WaitFor` shouldn't close the signal (only writers in the concurrency model should close the channel).

Figured a bit of code copying should be fine.